clients/xedit/xedit.man
.TH XEDIT 1 "26 October 1988" "X Version 11"
.SH NAME
xedit - simple text editor for X
.SH SYNTAX
\fBxedit\fP [ \fI-toolkitoption\fP ...] [ filename ]
.SH OPTIONS
.I Xedit
accepts all of the standard X Toolkit command line
options (see \fIX(1)\fP), plus:
.TP 8
.I filename
Specifies the file that is to be loaded during start-up.
This is the file which will be edited.
If a file is not specified,
.I xedit
lets you load a file or create a new file after it has started up.
.SH DESCRIPTION
.I Xedit
provides a window consisting of the following three areas:
.IP "Commands Menu" 25
Lists editing commands (for example, \fBUndo\fP or \fBSearch\fP).
.IP "Message Window"
Displays
.I xedit
messages.
In addition, this window can be used as a scratch pad.
.IP "Edit Window"
Displays the text of the file that you are editing or creating.
.SH COMMANDS
.IP "Quit" 25
Quits the current editing session.
If any changes have not been saved,
.I xedit 
displays a warning message and allows you to save the file.
.IP "Save"
Stores a copy of the original, unedited file in \fIfile\fP.BAK. 
Then, overwrites the original file with the edited contents.
.IP "Edit"
Allows the text displayed in the Edit window to be edited.
.IP "Load"
Loads the specified file and displays it in the Edit window.
.IP "Undo"
Undoes the last edit only.
.IP "More"
Undoes each edit previous to the last edit,
which must first be undone with the 
.B Undo
command.
.IP "Jump"
Advances the cursor from the beginning of the file to the text line
that corresponds to the selected line number.
.IP "<<"
Searches from the cursor back to the beginning of the file for the string
entered in the Search input box.
If you do not enter a string in the Search input box,
.I xedit
automatically copies the last string that you selected from any
X application into the Search input box and searches for that string.
.IP "Search >>"
Searches from the cursor forward to the end of the file for the string
entered in the search input box.
If you do not enter a string in the Search input box,
.I xedit
automatically copies the last string that you selected from any
X application into the Search input box and searches for that string.
.IP "Replace"
Replaces the last searched-for string
with the string specified in the Replace input box.
If no string has been previously searched for,
searches from the insert cursor to the end of the file 
for the next occurrence of the search string and highlights it.
.IP "All"
Repositions the cursor at the beginning of the file and replaces all 
occurrences of the search string with the string specified in the 
Replace input box.
.SH X DEFAULTS
For
.I xedit,
the available class identifiers are:
.sp
.nf
ButtonBox
Command
Scrollbar
Text   
.fi
.PP
For
.I xedit,
the available name identifiers are:
.sp
.nf
All
Edit
EditWindow
Jump
Load
MessageWindow
More
Quit
Replace
Save
Undo
xedit
.fi
.sp
.LP
For
.I xedit,
the available resources are:
.IP EnableBackups 22
Specifies that,
when edits made to an existing file are saved,
.I xedit
is to copy the original version of that file to \fIfile\fP.BAK 
before it saves the changes.
If the value of this option is specified as off, 
a backup file is not created.
.IP background
Specifies the background color to be displayed in command buttons.
The default is white.
.IP border
Specifies the border color of the 
.I xedit
window.
.IP borderWidth
Specifies the border width, in pixels, of the 
.I xedit
window.
.IP font
Specifies the font displayed in the
.I xedit
window.
.IP foreground
Specifies the foreground color of the
.I xedit
window.
The default is black.
.IP geometry
Specifies the geometry (window size and screen location)
to be used as the default for the 
.I xedit
window.
For information about the format of the geometry specification,
see \fIX(1)\fP.
.IP internalHeight
Specifies the internal horizontal padding (spacing between text and button
border) for command buttons.
.IP internalWidth
Specifies the internal vertical padding (spacing between text and button
border) for command buttons.
.SH KEY BINDINGS
.LP
Each specification included in the \fI.XtActions\fP file
modifies a key setting
for the editor that
.I xedit
uses.
When defining key specifications,
you must use the following resource specification:
.EX
text.EventBindings:     .XtActions
.EE
.LP
Each key specification assigns an editor command to a named
key and/or mouse combination and has the format:
.sp
.DS
.TA .5i 1.5i
.ta .5i 1.5i
\fIkey\fP\^:	\fIfunction\fP
.DE
.sp
.IP \fIkey\fP 15
Specifies the key or mouse button that is used to invoke the named function.
.IP \fIfunction\fP
Specifies the function to be invoked when the named key is pressed.
.LP
For more information about specifications in the
.I .XtActions
file, see \fIX(1)\fP.
.SH FILES
~/.XtActions
.br
/usr/lib/X11/.XtActions
.SH SEE ALSO
X(1), xrdb(1)
.SH RESTRICTIONS
Large numbers of certain edit functions (for example, Undo or More)
tend to degrade performance over time.
If there is a noticeable decrease in response time,
save and reload the file.
.SH BUGS
It is not clear how to select a line number for the \fIJump\fP command.
.PP
The string searches don't work properly.
.SH COPYRIGHT
Copyright 1988, Digital Equipment Corporation.


CHANGES
The following changes have been made to this directory since R2:

     o  substantially improved the X man page.
     o  added README file with quick instructions for those in the fast lane.
     o  added RELEASENOTES file giving overview of what has changed and where
	to look for more information.
     o  added STANDARDS file describing what is and is not part of the X
	Standard.
     o  removed support for automatically building contrib directory from
	the Imakefile since it won't work completely on any single platform.
     o  moved rgb directory here from util.
     o  added some misc. utility targets to the top-level Makefile.


clients/xedit/xedit.c
#ifndef lint
static char rcs_id[] = "$XConsortium: xedit.c,v 1.16 88/10/18 13:30:41 swick Exp $";
#endif

/*
 *			  COPYRIGHT 1987
 *		   DIGITAL EQUIPMENT CORPORATION
 *		       MAYNARD, MASSACHUSETTS
 *			ALL RIGHTS RESERVED.
 *
 * THE INFORMATION IN THIS SOFTWARE IS SUBJECT TO CHANGE WITHOUT NOTICE AND
 * SHOULD NOT BE CONSTRUED AS A COMMITMENT BY DIGITAL EQUIPMENT CORPORATION.
 * DIGITAL MAKES NO REPRESENTATIONS ABOUT THE SUITABILITY OF THIS SOFTWARE FOR
 * ANY PURPOSE.  IT IS SUPPLIED "AS IS" WITHOUT EXPRESS OR IMPLIED WARRANTY.
 *
 * IF THE SOFTWARE IS MODIFIED IN A MANNER CREATING DERIVATIVE COPYRIGHT RIGHTS,
 * APPROPRIATE LEGENDS MAY BE PLACED ON THE DERIVATIVE WORK IN ADDITION TO THAT
 * SET FORTH ABOVE.
 *
 *
 * Permission to use, copy, modify, and distribute this software and its
 * documentation for any purpose and without fee is hereby granted, provided
 * that the above copyright notice appear in all copies and that both that
 * copyright notice and this permission notice appear in supporting documentation,
 * and that the name of Digital Equipment Corporation not be used in advertising
 * or publicity pertaining to distribution of the software without specific, 
 * written prior permission.
 */

#include "xedit.h"
#include <X11/Shell.h>

#ifndef DEFWIDTH
#define DEFWIDTH 500
#endif

#ifndef DEFHEIGHT
#define DEFHEIGHT 700
#endif

int Editable;
int saved;
int backedup;
int lastChangeNumber;
char *displayname;
char *filename;
char *loadedfile;
char *savedfile;
char *searchstring;
char *replacestring;

Widget toplevel;
Widget outer;
Widget Row1;
Widget Row2;
Widget fileBox;
Widget replaceBox;
Widget searchBox;
Widget textwindow;
Widget messwidget;
Widget undoBox;
Widget labelwindow;
		
Widget quitbutton;
Widget loadbutton;
Widget savebutton;
Widget editbutton;
Widget filenamewindow;

Widget searchstringwindow;
Widget replacestringwindow;
Widget qbutton;
Widget ubutton;
Widget umbutton;
Widget searchbutton;
Widget lsearchbutton;
Widget replacebutton;
Widget replaceallbutton;
Widget jumpbutton;

Display *CurDpy;
XtTextSource source, asource, dsource, psource, messsource;
extern DoQ();


makeButtonsAndBoxes()
{
  int boxHeight;
  static Arg TextArgs[] = {
	{ XtNtextSource, NULL },
	{ XtNtextSink, NULL },
	{ XtNtextOptions , (XtArgVal)(scrollVertical | wordBreak) }, 
  };
  static Arg MessArgs[] = {
	{ XtNtextSource, NULL },
	{ XtNtextSink, NULL },
	{ XtNtextOptions, (XtArgVal)(scrollVertical | wordBreak) }, 
  };
  static Arg labelArgs[] = {
	{ XtNjustify,	 (XtArgVal)XtJustifyCenter },
	{ XtNlabel, 	(XtArgVal)"no file yet" },
  };
    outer = XtCreateManagedWidget( "vpaned", vPanedWidgetClass, toplevel,
                                  NULL, 0 );
    XtPanedSetRefigureMode(outer, FALSE);

/*    XtPanedAllowResize(outer, TRUE); */
    Row1 = XtCreateManagedWidget("row1", boxWidgetClass, outer, NULL,0);
        quitbutton = makeCommandButton(Row1, "Quit", DoQuit);
        savebutton = makeCommandButton(Row1, "Save", DoSave);
        editbutton = makeCommandButton(Row1, "Edit", DoEdit);
/*        qbutton = makeCommandButton(Row1, "q", DoQ);   */
        loadbutton = makeCommandButton(Row1, "Load", DoLoad);
        filenamewindow = makeStringBox(Row1, filename,110); 
        ubutton = makeCommandButton(Row1, "Undo", DoUndo);
        umbutton = makeCommandButton(Row1, "More", DoUndoMore);
        jumpbutton = makeCommandButton(Row1, "Jump", DoJump);
/*        linebutton = makeCommandButton(Row1, "Line?", DoLine);  */
    Row2 = XtCreateManagedWidget("row2", boxWidgetClass, outer, NULL,0);
        lsearchbutton = makeCommandButton(Row2, "<< ", DoSearchLeft);
        searchbutton = makeCommandButton(Row2,"Search >>",DoSearchRight); 
        searchstringwindow = makeStringBox(Row2, searchstring, 120); 
        replacebutton= makeCommandButton(Row2, "Replace", DoReplaceOne);
        replaceallbutton = makeCommandButton(Row2, "All", DoReplaceAll);
        replacestringwindow = makeStringBox(Row2, replacestring, 120); 
    {
    int _Min, _Max;
    Dimension boxHeight;
    Arg getargs[1];
    XtSetArg(getargs[0], XtNheight, &boxHeight);
    XtGetValues(quitbutton, getargs, XtNumber(getargs));
    boxHeight += 8;
    XtPanedSetMinMax((Widget) Row1, (int)boxHeight, (int)boxHeight);
    XtPanedSetMinMax((Widget) Row2, (int)boxHeight, (int)boxHeight);

    }
    messsource = PseudoDiskSourceCreate("");
    MessArgs[0].value = (XtArgVal)messsource;
    MessArgs[1].value = (XtArgVal)XtAsciiSinkCreate(outer, NULL, 0);
    messwidget =  XtCreateManagedWidget("messageWindow", textWidgetClass,
		outer, MessArgs, XtNumber(MessArgs));
    XtPanedSetMinMax((Widget) messwidget, 40, 40);

    labelwindow = XtCreateManagedWidget("labelWindow",labelWidgetClass, 
		outer, labelArgs, XtNumber(labelArgs)); 
    XtPanedSetMinMax((Widget)labelwindow, 14, 14);
    TextArgs[0].value = (XtArgVal)source;
    TextArgs[1].value = (XtArgVal)XtAsciiSinkCreate(outer, NULL, 0);
    textwindow =  XtCreateManagedWidget("editWindow", textWidgetClass, 
		outer, TextArgs, XtNumber(TextArgs));
    XtPanedSetRefigureMode(outer, TRUE); 
}

struct _app_resources app_resources;

#define offset(field) XtOffset(struct _app_resources*, field)
static XtResource resources[] = {
   {"editInPlace", "EditInPlace", XtRBoolean, sizeof(int),
         offset(editInPlace), XtRString, "False"},
   {"enableBackups", "EnableBackups", XtRBoolean, sizeof(int),
         offset(enableBackups), XtRString, "FALSE"},
   {"backupNamePrefix", "BackupNamePrefix", XtRString, sizeof(char *),
         offset(backupNamePrefix),XtRString, ""},
   {"backupNameSuffix", "BackupNameSuffix", XtRString, sizeof(char *),
         offset(backupNameSuffix),XtRString, ".BAK"}
};
#undef offset

main(argc, argv)
  int argc;
  char **argv;
{
    static Dimension width, height; /* init'd to 0 */
    Arg args[3];
    backedup = 0;
    saved = 0;
    filename = malloc(1000);
    searchstring = malloc(1000);
    replacestring = malloc(1000);
    toplevel = XtInitialize( "xedit", "Xedit", NULL, 0, &argc, argv);
    XtSetArg(args[0], XtNwidth, &width);
    XtSetArg(args[1], XtNheight, &height);
    XtGetValues(toplevel, args, XtNumber(args));
    if (width == 0) width = DEFWIDTH;
    if (height == 0) height = DEFHEIGHT;
    XtSetArg(args[0], XtNwidth, width);
    XtSetArg(args[1], XtNheight, height);
    XtSetArg(args[2], XtNinput, True);
    XtSetValues(toplevel, args, (Cardinal)3);

    XtGetApplicationResources(toplevel, &app_resources, resources,
                                  XtNumber(resources),NULL, 0);
    CurDpy = XtDisplay(toplevel);
    dsource = PseudoDiskSourceCreate(""); 
    asource = PseudoDiskSourceCreate("");
    source = CreatePSource(dsource, asource);  
    if(argv[1]) strcpy(filename, argv[1]);
    makeButtonsAndBoxes();
    XtRealizeWidget(toplevel);
    XDefineCursor(XtDisplay(toplevel),XtWindow(toplevel),
         XCreateFontCursor( XtDisplay(toplevel), XC_left_ptr));
    if (strlen(filename)) DoLoad();
    XtMainLoop();
}

